sim/battle.ts

```diff
@@ -479,10 +509,26 @@ export class Battle {
 	}
 
 	/** The entire event system revolves around this function and runEvent. */
+	singleEvent<Event extends `${EventNamePrefix}${EventName}`, MainEffect extends Effect>(
```

Comment:
What's the use case for allowing something like `singleEvent('SourceHit', ...)`? I feel like I've used this as a hack before (probably years ago) but it probably shouldn't normally be allowed.

Reply:
Aha! It appears `onMaybeTrapPokemon` exists as a method on the interface but is unused, only `onFoeMaybeTrapPokemon` is used in practice. So, we could probably just rename the event to `MaybeTrapFoe` or similar so it doesn't get confused with an actual prefix.

Reply:
The main call is here:

https://github.com/smogon/pokemon-showdown/blob/master/sim/battle.ts#L1495

What you pointed out is an additional call for the "You _might_ be trapped, so you won't be able to cancel a switch" check. (Which doesn't make sense for AAA, Hackmons etc. but there's no check for obtainable abilities in the ruleset.)